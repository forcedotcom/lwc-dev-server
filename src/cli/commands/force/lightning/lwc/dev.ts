import { flags, SfdxCommand } from '@salesforce/command';
import { Messages, SfdxError } from '@salesforce/core';
import { AnyJson, JsonArray, JsonMap } from '@salesforce/ts-types';
import * as http from 'http';
import Project from '../../../../../common/Project';
import SfdxConfiguration from '../../../../../user/SfdxConfiguration';
import LocalDevServer from '../../../../../server/LocalDevServer';
import debugLogger from 'debug';

const debug = debugLogger('localdevserver');

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('lwc-dev-server', 'dev');

export default class Dev extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lighting:lwc:dev --open myComponent`
    ];

    public static args = [{ open: 'file', name: 'lwc-dev' }];

    protected static flagsConfig = {
        open: flags.string({
            char: 'o',
            description: messages.getMessage('openFlagDescription')
        }),
        port: flags.integer({
            char: 'p',
            description: messages.getMessage('portFlagDescription')
        })
    };

    // Comment this out if your command does not require an org username
    protected static requiresUsername = true;

    // Comment this out if your command does not support a hub org username
    protected static supportsDevhubUsername = true;

    // Set this to true if your command requires a project workspace; 'requiresProject' is false by default
    protected static requiresProject = true;

    public async run(): Promise<AnyJson> {
        const componentName = this.flags.open;
        const port = this.flags.port || 3333;

        if (!this.org) {
            this.ux.log('org was undefined, an org is required.');
            return { org: typeof this.org };
        }

        if (!this.project) {
            this.ux.log(
                'project is undefined. this must be run from a sfdx worksapce'
            );
            return { project: typeof this.project };
        }

        // this.org is guaranteed because requiresUsername=true, as opposed to supportsUsername
        const conn = this.org.getConnection();

        // Highest level API is always last
        const api_version = await conn.retrieveMaxApiVersion();

        debug(
            `You appear to be running on a Salesforce instance that can support up to API level ${api_version}`
        );

        if (componentName) {
            debug(`You wanted to open this component: ${componentName}`);
        }

        // TODO resolve location of `componentName`, ensure directory structure is imported / compiled

        // TODO check if it's already running on the port first

        // custom onProxyReq function to inject into Talon's proxy
        // this will insert the Authorization header to have the requests be authenticated
        const onProxyReq = function(
            proxyReq: http.ClientRequest,
            req: http.IncomingMessage,
            res: http.ServerResponse
        ) {
            proxyReq.setHeader('Authorization', `Bearer ${conn.accessToken}`);
            // req.headers.Cookie = `sid=${sid_cookie}`;
        };

        const project = new Project(this.project.getPath());

        const sfdxConfiguration = new SfdxConfiguration(project);
        sfdxConfiguration.api_version = api_version;
        sfdxConfiguration.endpoint = conn.instanceUrl;
        sfdxConfiguration.onProxyReq = onProxyReq;
        sfdxConfiguration.port = port;
        sfdxConfiguration.namespace = <string>(
            (await this.project.resolveProjectConfig()).namespace
        );

        project.sfdxConfiguration = sfdxConfiguration;

        const retValue = {
            orgId: this.org.getOrgId(),
            api_version: sfdxConfiguration.api_version,
            endpoint: sfdxConfiguration.endpoint,
            onProxyReq: JSON.stringify(sfdxConfiguration.onProxyReq),
            componentName,
            port,
            token: conn.accessToken
        };
        debug(JSON.stringify(retValue));

        // Start local dev server
        // TODO pass in component to open & open browser
        new LocalDevServer().start(project);

        return retValue;
    }
}
