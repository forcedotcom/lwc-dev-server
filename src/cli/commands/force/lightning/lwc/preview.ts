#!/usr/bin/env ts-node
import { flags, SfdxCommand } from '@salesforce/command';
import { Messages, SfdxError } from '@salesforce/core';
import { CommandLineUtils } from '../../../../../common/Common';
import { Launcher } from '../../../../../common/Requirements';
import { LaunchIOS } from '../../../../../common/LaunchIOS';
import { LaunchAndroid } from '../../../../../common/LaunchAndroid';
import { DefaultLWCMobileConfig } from '../../../../../common/Config';

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('@salesforce/lwc-dev-server', 'preview');

export default class Preview extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lightning:lwc:preview -p iOS -t LWCSim2 -f http://localhost:3333`,
        `$ sfdx force:lightning:lwc:preview -p Android -t LWCEmu2 -f http://localhost:3333`
    ];

    public static args = [{ name: 'file' }];

    protected static flagsConfig = {
        // flag with a value (-n, --name=VALUE)
        platform: flags.string({
            char: 'p',
            description: messages.getMessage('platformFlagDescription')
        }),
        path: flags.string({
            char: 'f',
            description: messages.getMessage('pathFlagDescription')
        }),
        target: flags.string({
            char: 't',
            description: messages.getMessage('targetFlagDescription')
        })
    };

    // Comment this out if your command does not require an org username
    protected static requiresUsername = false;

    // Comment this out if your command does not support a hub org username
    protected static supportsDevhubUsername = false;

    // Set this to true if your command requires a project workspace; 'requiresProject' is false by default
    protected static requiresProject = false;

    public async run(): Promise<any> {
        let launcher: Launcher;
        let url: string = '';

        if (this.flags.platform === undefined) {
            return new Promise<any>((resolve, reject) => reject(false));
        }
        //FIXME: Invoke setup and use detected settings!
        if (CommandLineUtils.platformFlagIsIOS(this.flags.platform)) {
            launcher = new LaunchIOS(this.flags.target, 'iPhone-X', 'iOS-13-3');
            url = this.flags.path;
        } else if (
            CommandLineUtils.platformFlagIsAndroid(this.flags.platform)
        ) {
            launcher = new LaunchAndroid(
                this.flags.target,
                'pixel',
                'android-28'
            );
            url = this.flags.path;
        }

        // if (this.flags.platform.toLowerCase() == "ios") {
        //   launcher = new LaunchIOS(this.flags.target, "iPhone-X","iOS-13-3");
        //   url = this.flags.path;
        // } else if (this.flags.platform.toLowerCase() == "android") {
        //   launcher = new LaunchAndroid(this.flags.target,"pixel","android-28");
        //   url = this.flags.path;
        // }
        // return launcher.launchNativeBrowser(url);
        return new Promise<any>((resolve, reject) => {
            if (launcher == null) {
                reject('Could not launch target');
                return;
            }
            resolve(launcher.launchNativeBrowser(url));
        });
    }
}
