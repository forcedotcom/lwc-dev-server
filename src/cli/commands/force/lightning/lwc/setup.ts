#!/usr/bin/env ts-node

import { flags, SfdxCommand } from '@salesforce/command';
import { Messages, SfdxError } from '@salesforce/core';
import { CommandLineUtils } from '../../../../../common/Common';
import {
    RequirementList,
    SetupTestResult
} from '../../../../../common/Requirements';
import { AndroidEnvironmentSetup } from '../../../../../common/AndroidEnvironmentSetup';
import { IOSEnvironmentSetup } from '../../../../../common/IOSEnvironmentSetup';

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('@salesforce/lwc-dev-server', 'preview');

export default class Setup extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lightning:lwc:setup -p iOS`,
        `$ sfdx force:lightning:lwc:setup -p Android`
    ];

    protected static flagsConfig = {
        // flag with a value (-n, --name=VALUE)
        platform: flags.string({
            char: 'p',
            description: messages.getMessage('platformFlagDescription')
        })
    };

    public async run(): Promise<SetupTestResult> {
        let setup: RequirementList;
        //let setupResult: SetupTestResult = new  SetupTestResult();

        if (this.flags.platform === undefined) {
            return new Promise((resolve, reject) => {
                reject({ hasMetAllRequirements: false, tests: [] });
            });
        }

        if (CommandLineUtils.platformFlagIsIOS(this.flags.platform)) {
            setup = new IOSEnvironmentSetup();
        } else {
            setup = new AndroidEnvironmentSetup();
        }
        return setup.executeSetup();
    }
}
