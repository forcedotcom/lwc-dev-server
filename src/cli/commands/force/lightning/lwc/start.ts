import { flags, SfdxCommand } from '@salesforce/command';
import { Messages } from '@salesforce/core';
import { AnyJson } from '@salesforce/ts-types';
import * as http from 'http';
import Project from '../../../../../common/Project';
import LocalDevServer from '../../../../../server/LocalDevServer';
import debugLogger from 'debug';
import colors from 'colors';

const debug = debugLogger('localdevserver');

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('@salesforce/lwc-dev-server', 'start');

export default class Start extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lighting:lwc:start`,
        `$ sfdx force:lighting:lwc:start --port 3000`
    ];

    public static args = [];

    protected static flagsConfig = {
        port: flags.integer({
            char: 'p',
            description: messages.getMessage('portFlagDescription')
        })
    };

    // Comment this out if your command does not support specifying an org username
    protected static supportsUsername = true;

    // Comment this out if your command does not support a dev hub org username
    protected static supportsDevhubUsername = true;

    // Set this to true if your command requires a project workspace; 'requiresProject' is false by default
    protected static requiresProject = true;

    public async run(): Promise<AnyJson> {
        const defaultdevhubusername = this.configAggregator.getPropertyValue(
            'defaultdevhubusername'
        ) as string;

        let port: number;
        if (this.flags.port !== undefined && this.flags.port !== null) {
            port = this.flags.port;
        } else {
            port = 3333;
        }

        if (!this.org) {
            // This you DO need.
            // We require this right now for proxying and api version.
            // If we do not have an org, we can still function.
            //  - We should disable the proxying and then allow specification of the version
            //    if you don't authenticate.
            //

            const targetusername = this.flags.targetusername;
            if (targetusername) {
                this.reportStatus(
                    colors.green(defaultdevhubusername),
                    colors.red(
                        `${targetusername} - ${messages.getMessage(
                            'error:invalidscratchorgusername'
                        )}`
                    )
                );
            } else {
                const configuredusername = this.configAggregator.getPropertyValue(
                    'defaultusername'
                );
                this.reportStatus(
                    colors.green(defaultdevhubusername),
                    colors.red(
                        `${configuredusername} - ${messages.getMessage(
                            'error:noscratchorg'
                        )}`
                    )
                );
            }

            return { org: typeof this.org };
        }

        // Sfdx validates this before we have a chance to, this appears to be
        // a "just in case" condition so reporting the same error they do.
        if (!this.project) {
            this.ux.error(messages.getMessage('error:noproject'));
            return { project: typeof this.project };
        }

        this.ux.log(colors.gray(messages.getMessage('legal:cliusage')));

        // this.org is guaranteed because requiresUsername=true, as opposed to supportsUsername
        const conn = this.org.getConnection();

        // Highest level API is always last
        const api_version = await conn.retrieveMaxApiVersion();

        const orgusername = this.org.getUsername() || '';
        try {
            // currently something in sfdx is resulting in an unhandled
            // promise rejection, instead of surfacing that rejection
            // through the returned promise. @W-6723813
            process.on('unhandledRejection', reason => {
                if (
                    reason &&
                    // @ts-ignore
                    reason.name &&
                    // @ts-ignore
                    reason.name === 'StatusCodeError'
                ) {
                    // ignore unhandled rejects during below refresh call
                    debug(`unhandledPromiseRejection: ${reason}`);
                } else {
                    this.ux.error(`unhandledPromiseRejection: ${reason}`);
                }
            });
            await this.org.refreshAuth();
        } catch (err) {
            this.reportStatus(
                colors.green(defaultdevhubusername),
                colors.red(
                    `${orgusername} - ${messages.getMessage(
                        'error:inactivescratchorg'
                    )}`
                ),
                colors.green(api_version)
            );
            return { error: err };
        }

        this.reportStatus(
            colors.green(defaultdevhubusername),
            colors.green(orgusername),
            colors.green(api_version)
        );

        const accessToken = conn.accessToken;

        // custom onProxyReq function to inject into Talon's proxy
        // this will insert the Authorization header to have the requests be authenticated
        const onProxyReq = function(
            proxyReq: http.ClientRequest,
            req: http.IncomingMessage,
            res: http.ServerResponse
        ) {
            proxyReq.setHeader('Authorization', `Bearer ${accessToken}`);
        };

        const project = new Project(this.project.getPath());

        project.configuration.api_version = api_version;
        project.configuration.endpoint = conn.instanceUrl;
        project.configuration.onProxyReq = onProxyReq;
        project.configuration.port = port;
        project.configuration.namespace = <string>(
            (await this.project.resolveProjectConfig()).namespace
        );

        const retValue = {
            orgId: this.org.getOrgId(),
            api_version: project.configuration.api_version,
            endpoint: project.configuration.endpoint,
            onProxyReq: JSON.stringify(project.configuration.onProxyReq),
            port,
            token: accessToken
        };
        debug(JSON.stringify({ ...retValue, token: undefined }));

        // Start local dev server
        new LocalDevServer().start(project, conn);

        return retValue;
    }

    private reportStatus(
        devHubOrg: string,
        scratchOrg: string,
        apiVersion?: string
    ) {
        if (apiVersion) {
            this.ux.log(`\
Starting LWC Local Development.
    Dev Hub Org: ${devHubOrg}
    Scratch Org: ${scratchOrg}
    Api Version: ${apiVersion}\
`);
        } else {
            this.ux.log(`\
Starting LWC Local Development.
    Dev Hub Org: ${devHubOrg}
    Scratch Org: ${scratchOrg}\
`);
        }
    }
}
