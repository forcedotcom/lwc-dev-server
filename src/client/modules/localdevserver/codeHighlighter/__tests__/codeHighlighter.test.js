import { createElement } from 'lwc';
import CodeHighlighter from 'localdevserver/codeHighlighter';

jest.useFakeTimers();

function createComponentUnderTest(props) {
    const el = createElement('localdevserver-code-highlighter', {
        is: CodeHighlighter
    });
    Object.assign(el, props);
    document.body.appendChild(el);
    return el;
}
const code = `
// this is sameple code
// forEach polyfill
Array.prototype.forEach = function(callback, thisArg) {
    if(typeof(callback) !== "function") {
        throw new TypeError(callback + " is not a function!");
    }
    var len = this.length;
    for(var i = 0; i < len; i++) {
        callback.call(thisArg, this[i], i, this)
    }
}`;
describe('localdevserver-code-highlighter', () => {
    it('renders - no code', async () => {
        const componentElement = createComponentUnderTest();
        jest.runAllTicks();
        expect(componentElement).toMatchSnapshot();
    });
    it('renders - code, no line info', async () => {
        const componentElement = createComponentUnderTest();
        jest.runAllTicks();
        componentElement.code = code;
        jest.runAllTicks();
        jest.runAllTimers();
        expect(componentElement).toMatchSnapshot();

        const pre = componentElement.shadowRoot.querySelector('pre');
        expect(pre.classList.contains('language-javascript')).toBe(true);
    });
    it('renders - code, with line info', async () => {
        const componentElement = createComponentUnderTest();
        jest.runAllTicks();
        componentElement.code = code;
        componentElement.lineOffset = 3;
        jest.runAllTicks();
        jest.runAllTimers();
        expect(componentElement).toMatchSnapshot();

        const pre = componentElement.shadowRoot.querySelector('pre');
        expect(pre.classList.contains('language-javascript')).toBe(true);
    });
});
