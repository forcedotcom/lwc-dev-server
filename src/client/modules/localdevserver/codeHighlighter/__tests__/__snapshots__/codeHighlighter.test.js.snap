// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localdevserver-code-highlighter renders - code, no line info 1`] = `
<localdevserver-code-highlighter>
  #shadow-root(open)
    <div>
      <pre
        class=" language-javascript"
      >
        <code
          class=" language-javascript"
        >
          

          <span
            class="token comment"
          >
            // this is sameple code
          </span>
          

          <span
            class="token comment"
          >
            // forEach polyfill
          </span>
          

          <span
            class="token class-name"
          >
            Array
          </span>
          <span
            class="token punctuation"
          >
            .
          </span>
          prototype
          <span
            class="token punctuation"
          >
            .
          </span>
          <span
            class="token function-variable function"
          >
            forEach
          </span>
           
          <span
            class="token operator"
          >
            =
          </span>
           
          <span
            class="token keyword"
          >
            function
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token parameter"
          >
            callback
            <span
              class="token punctuation"
            >
              ,
            </span>
             thisArg
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
           
          <span
            class="token punctuation"
          >
            {
          </span>
          
    
          <span
            class="token keyword"
          >
            if
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token keyword"
          >
            typeof
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          callback
          <span
            class="token punctuation"
          >
            )
          </span>
           
          <span
            class="token operator"
          >
            !==
          </span>
           
          <span
            class="token string"
          >
            "function"
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
           
          <span
            class="token punctuation"
          >
            {
          </span>
          
        
          <span
            class="token keyword"
          >
            throw
          </span>
           
          <span
            class="token keyword"
          >
            new
          </span>
           
          <span
            class="token class-name"
          >
            TypeError
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          callback 
          <span
            class="token operator"
          >
            +
          </span>
           
          <span
            class="token string"
          >
            " is not a function!"
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
          
    
          <span
            class="token punctuation"
          >
            }
          </span>
          
    
          <span
            class="token keyword"
          >
            var
          </span>
           len 
          <span
            class="token operator"
          >
            =
          </span>
           
          <span
            class="token keyword"
          >
            this
          </span>
          <span
            class="token punctuation"
          >
            .
          </span>
          length
          <span
            class="token punctuation"
          >
            ;
          </span>
          
    
          <span
            class="token keyword"
          >
            for
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token keyword"
          >
            var
          </span>
           i 
          <span
            class="token operator"
          >
            =
          </span>
           
          <span
            class="token number"
          >
            0
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
           i 
          <span
            class="token operator"
          >
            &lt;
          </span>
           len
          <span
            class="token punctuation"
          >
            ;
          </span>
           i
          <span
            class="token operator"
          >
            ++
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
           
          <span
            class="token punctuation"
          >
            {
          </span>
          
        
          <span
            class="token function"
          >
            callback
          </span>
          <span
            class="token punctuation"
          >
            .
          </span>
          <span
            class="token function"
          >
            call
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          thisArg
          <span
            class="token punctuation"
          >
            ,
          </span>
           
          <span
            class="token keyword"
          >
            this
          </span>
          <span
            class="token punctuation"
          >
            [
          </span>
          i
          <span
            class="token punctuation"
          >
            ]
          </span>
          <span
            class="token punctuation"
          >
            ,
          </span>
           i
          <span
            class="token punctuation"
          >
            ,
          </span>
           
          <span
            class="token keyword"
          >
            this
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
          
    
          <span
            class="token punctuation"
          >
            }
          </span>
          

          <span
            class="token punctuation"
          >
            }
          </span>
        </code>
      </pre>
    </div>
</localdevserver-code-highlighter>
`;

exports[`localdevserver-code-highlighter renders - code, with line info 1`] = `
<localdevserver-code-highlighter>
  #shadow-root(open)
    <div>
      <pre
        class="  language-javascript"
      >
        <code
          class="  language-javascript"
        >
          

          <span
            class="token comment"
          >
            // this is sameple code
          </span>
          

          <span
            class="token comment"
          >
            // forEach polyfill
          </span>
          

          <span
            class="token class-name"
          >
            Array
          </span>
          <span
            class="token punctuation"
          >
            .
          </span>
          prototype
          <span
            class="token punctuation"
          >
            .
          </span>
          <span
            class="token function-variable function"
          >
            forEach
          </span>
           
          <span
            class="token operator"
          >
            =
          </span>
           
          <span
            class="token keyword"
          >
            function
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token parameter"
          >
            callback
            <span
              class="token punctuation"
            >
              ,
            </span>
             thisArg
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
           
          <span
            class="token punctuation"
          >
            {
          </span>
          
    
          <span
            class="token keyword"
          >
            if
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token keyword"
          >
            typeof
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          callback
          <span
            class="token punctuation"
          >
            )
          </span>
           
          <span
            class="token operator"
          >
            !==
          </span>
           
          <span
            class="token string"
          >
            "function"
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
           
          <span
            class="token punctuation"
          >
            {
          </span>
          
        
          <span
            class="token keyword"
          >
            throw
          </span>
           
          <span
            class="token keyword"
          >
            new
          </span>
           
          <span
            class="token class-name"
          >
            TypeError
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          callback 
          <span
            class="token operator"
          >
            +
          </span>
           
          <span
            class="token string"
          >
            " is not a function!"
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
          
    
          <span
            class="token punctuation"
          >
            }
          </span>
          
    
          <span
            class="token keyword"
          >
            var
          </span>
           len 
          <span
            class="token operator"
          >
            =
          </span>
           
          <span
            class="token keyword"
          >
            this
          </span>
          <span
            class="token punctuation"
          >
            .
          </span>
          length
          <span
            class="token punctuation"
          >
            ;
          </span>
          
    
          <span
            class="token keyword"
          >
            for
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token keyword"
          >
            var
          </span>
           i 
          <span
            class="token operator"
          >
            =
          </span>
           
          <span
            class="token number"
          >
            0
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
           i 
          <span
            class="token operator"
          >
            &lt;
          </span>
           len
          <span
            class="token punctuation"
          >
            ;
          </span>
           i
          <span
            class="token operator"
          >
            ++
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
           
          <span
            class="token punctuation"
          >
            {
          </span>
          
        
          <span
            class="token function"
          >
            callback
          </span>
          <span
            class="token punctuation"
          >
            .
          </span>
          <span
            class="token function"
          >
            call
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          thisArg
          <span
            class="token punctuation"
          >
            ,
          </span>
           
          <span
            class="token keyword"
          >
            this
          </span>
          <span
            class="token punctuation"
          >
            [
          </span>
          i
          <span
            class="token punctuation"
          >
            ]
          </span>
          <span
            class="token punctuation"
          >
            ,
          </span>
           i
          <span
            class="token punctuation"
          >
            ,
          </span>
           
          <span
            class="token keyword"
          >
            this
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
          
    
          <span
            class="token punctuation"
          >
            }
          </span>
          

          <span
            class="token punctuation"
          >
            }
          </span>
        </code>
      </pre>
    </div>
</localdevserver-code-highlighter>
`;

exports[`localdevserver-code-highlighter renders - no code 1`] = `
<localdevserver-code-highlighter>
  #shadow-root(open)
    <div />
</localdevserver-code-highlighter>
`;
